### PR TITLE
HIP-584: Refactor Precompiles

### PR DESCRIPTION
**Pull request by 0xivanov**
**Description**:

Refactor precompiles to match services:

* Move `SyntheticTxnFactory` to `AbstractWritePrecompile` 
* Remove `store` from the run method parameters
* Extract `store` from `MessageFrame`
* Add tests for `SyntheticTxnFactory`

**Related issue(s)**:

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java [hunk 1]
```diff
@@ -96,7 +95,7 @@ public class TransferPrecompile extends AbstractWritePrecompile {
     private int numLazyCreates;
 
     public TransferPrecompile(PrecompilePricingUtils pricingUtils, int functionId, boolean isLazyCreationEnabled) {
-        super(pricingUtils);
+        super(pricingUtils, null);
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:98]: Enhance the constructor to accept SyntheticTxnFactory

Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:98]: Fixed

Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java]: Fixed

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**